Sources/AppcuesKit/Presentation/Public/AppcuesElementTargeting.swift

```diff
@@ -0,0 +1,35 @@
+//
+//  AppcuesElementTargeting.swift
+//  AppcuesKit
+//
+//  Created by James Ellis on 3/20/23.
+//  Copyright © 2023 Appcues. All rights reserved.
+//
+
+import Foundation
+
+/// A type that describes an element targeting strategy for views in the application UI.
+///
+/// The SDK provides a default implementation for element targeting, but it can be overridden at initialzation of the SDK
+/// using the `Appcues.Config` option for `elementTargeting`.
+@objc
+public protocol AppcuesElementTargeting {
+    /// Capture the layout hierarchy in the currently rendered screen of the application.
+    /// - Returns: Root view element for the current screen, or `nil` if not available. The view element
+    ///            contains sub-views recursively in the `children` property.
+    func captureLayout() -> AppcuesViewElement?
+
+    /// Create and return a selector implementation from configuration properties.
+    /// - Parameter properties: Key-value collection of properties to identify a given view element.
+    /// - Returns: The `AppcuesElementSelector` implementation for this element targeting stragegy, based on
+    ///            the given properties. If no applicable properties are found, the return value should be `nil`.
+    func inflateSelector(from properties: [String: String]) -> AppcuesElementSelector?
+}
+
+internal extension AppcuesElementTargeting {
+    // Default implementation that looks at the current view capture and finds any matches.
+    // Used during targeted element experiences to find the position in the UI for the experience.
+    func findMatches(for selector: AppcuesElementSelector) -> [AppcuesViewElement] {
+        return captureLayout()?.viewsMatchingSelector(selector) ?? []
+    }
```

Comment:
Calling out that this is a bit less peformant than before. It now traverses the view tree once to capture the layout, then walks that tree a second time to find the views matching the selector. This could be combined into a single traversal of the view trying to find selector matches, but that would shift responsibility into the AppcuesElementTargeting implementation, rather than just using the layout capture it already provides in the internal spot here. I went this route to start, to lessen the implementation burden of any custom targeting strategy.